src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,810 @@
+  * **On Serialize**: When writing complex types, the serializer also writes them metadata ($id, $values and $ref) properties in order re-use them by writing a reference to the object or array.
+  * **On Deserialize**: While the other options show no effect on Deserialization, `Preserve` does affect its behavior with the following: Metadata will be expected (although is not mandatory) and the deserializer will try to understand it. 
+
+* **Ignore**:
+  * **On Serialize**: Ignores (skips writing) the property/element where the reference loop is detected.
+  * **On Deserialize**: Metadata properties will not be consumed, therefore they will be treated as regular properties that can map to a real property using `JsonPropertyName` or be added to the `JsonExtensionData` dictionary.
+
+For System.Text.Json, the goal is to stick to the same *metadata* semantics for preserve from Newtonsoft.Json and provide a similar usage in `JsonSerializerOptions` that encompasses the needed options (i.e. provide reference preservation).
+
+This API is exposing the `ReferenceHandling` property as a class, to be extensible in the future; and provide built-in static instances of `Default` and `Preserve` that are useful to enable the most common behaviors by just setting those in `JsonSerializerOptions.ReferenceHandling`.
+
+With `ReferenceHandling` being a class, we can exclude things that, as of now, we are not sure are required and add them later based on customer feedback. For example, the `Object` and `Array` granularity of `Newtonsoft.Json's` `PreserveReferencesHandling` or the `ReferenceLoopHandling.Ignore` option.
+
+## Compatibility
+
+The next table show the combination of Newtonsoft's **ReferenceLoopHandling** and **PreserveReferencesHandling** and how to get its equivalent on System.Text.Json's *ReferenceHandling*:
+
+|       RLH\PRH |      None |              All |          Objects |           Arrays |
+|--------------:|----------:|-----------------:|-----------------:|-----------------:|
+|     **Error** | *Default* | future (overlap) | future (overlap) | future (overlap) |
+|    **Ignore** |  *Ignore* | future (overlap) | future (overlap) | future (overlap) |
+| **Serialize** |    future |       *Preserve* |           future |           future |
+
+Notes: 
+* Newtonsoft's `MetadataPropertyHandling.ReadAhead` will not be supported in this first effort.
+* `Objects` and `Arrays` granularity may apply to both, Serialization and Deserialization.
+* (overlap) means that preserve references will co-exist with Reference Loop Handling and we will need to define how to resolve that (On Newtonsoft.Json, `PreserveReferencesHandling` takes precedence); see [example](#using-a-custom-referencehandling-to-show-possible-future-usage).
+
+
+# Examples
+
+Having the following class:
+```cs
+class Employee 
+{ 
+    string Name { get; set; }
+    Employee Manager { get; set; }
+    List<Employee> Subordinates { get; set; }
+}
+```
+
+## Using Default on Serialize
```

Comment:
Can you add an example for default deserialize (where you have metadata in the payload)?

Reply:
With the point of show how properties will add to `JsonExtensionData`, in case there is any?

Reply:
Yes and also if the .NET object uses the reserved keywords as property names (like `$id`), they will get matched with it and the value deserialized.

Reply:
Done, see https://github.com/dotnet/runtime/pull/354/files#diff-862e8417d479d30d1ade2dd9acf21333R114